pageserver/src/remote_storage/storage_sync.rs

```diff
@@ -240,82 +376,22 @@ fn storage_sync_loop<P: std::fmt::Debug, S: 'static + RemoteStorage<StoragePath
             }
         })
         .cloned()
-        .map(SyncTask::UrgentDownload)
-        .collect::<Vec<_>>();
-    info!(
-        "Will download {} timelines to restore state",
-        urgent_downloads.len()
-    );
-    let mut accessor = SYNC_QUEUE.lock().unwrap();
-    accessor.extend(urgent_downloads.into_iter());
-    drop(accessor);
+        .map(SyncKind::Download)
+        .for_each(|sync_kind| SYNC_QUEUE.push(SyncTask::new(0, sync_kind)));
+}
 
-    while !crate::tenant_mgr::shutdown_requested() {
-        let mut queue_accessor = SYNC_QUEUE.lock().unwrap();
-        let next_task = queue_accessor.pop();
-        let remaining_queue_length = queue_accessor.len();
-        drop(queue_accessor);
+fn get_next_tasks(max_concurrent_sync: NonZeroUsize) -> Vec<SyncTask> {
+    let mut max_concurrent_sync = max_concurrent_sync.get();
+    let mut tasks = Vec::with_capacity(max_concurrent_sync);
+    while let Some(new_task) = SYNC_QUEUE.pop() {
```

Comment:
Too bad none of these queues have batch reads

Reply:
Yes, https://docs.rs/crossbeam-deque/0.8.1/crossbeam_deque/ is intended to be for that, but it's a bit unusual in api-wise.